core/src/main/java/org/apache/iceberg/SnapshotOperations.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+
+/**
+ * SnapshotOperations abstracts access to snapshots for table metadata. This allows a subset of
+ * snapshots to be loaded initially for operations that do not require all snapshots to be present
+ * (e.g. table scans of the current snapshot).
+ *
+ * <p>In the event that snapshots need to be accessed, the provided supplier will be invoked, which
+ * must provide all addressable snapshots for the table.
+ */
+class SnapshotOperations implements Serializable {
+
+  private List<Snapshot> snapshots;
+  private Supplier<List<Snapshot>> snapshotsSupplier;
+  private Map<Long, Snapshot> snapshotsById;
+  private Map<String, SnapshotRef> refs;
+
+  SnapshotOperations() {
+    snapshots = Lists.newArrayList();
+    snapshotsById = Maps.newHashMap();
+    refs = Maps.newHashMap();
+  }
+
+  SnapshotOperations(List<Snapshot> snapshots, Map<String, SnapshotRef> refs) {
+    this(snapshots, null, refs);
+  }
+
+  SnapshotOperations(SnapshotOperations fromCopy) {
```

Comment:
I think it should be `toCopy` since the passed in one is the one we're copying?

Reply:
Two votes for `toCopy`, so I switched it.

Reply:
We use `toCopy` in other places, so this is a good change.